core/src/main/java/com/linecorp/armeria/internal/DefaultAttributeMap.java

```diff
@@ -155,19 +178,34 @@
                 }
                 curr = curr.next;
             }
-            return false;
+            return hasAttrInRoot(key);
         }
     }
 
-    /**
-     * Returns all {@link Attribute}s this map contains.
-     */
+    private <T> boolean hasAttrInRoot(AttributeKey<T> key) {
+        if (root != null) {
+            return root.hasAttr(key);
+        }
+        return false;
+    }
+
+    @Override
     public Iterator<Attribute<?>> attrs() {
         final AtomicReferenceArray<DefaultAttribute<?>> attributes = this.attributes;
         if (attributes == null) {
-            return Collections.emptyIterator();
+            return root != null ? root.attrs() : Collections.emptyIterator();
         }
 
+        final Iterator<Attribute<?>> ownAttrsIt = new IteratorImpl(attributes);
+        return root != null ? new ConcatenatedIteratorImpl(ownAttrsIt, root.attrs(), this)
```

Comment:
- Same here. When `root.attrs()` is involved, we need to make sure any changes do not go to the root but to `this`.
- Instead of passing `this`, you could just make `ConcatenatedIteratorImpl` non-static.
- Maybe not worth, but we may want to invoke `root.attrs()` only after the iteration of `ownAttrsIt` is over.

Reply:
All fixed except the last one. If we delay it we can:
- save one `findNext()` call
- `rootIt` can have attributes instead of being an `emptyIterator()` if there was an insertion before the iteration of `ownAttrsIt` is over. (which is rare)

I'm not sure if these have the advantage over having the final `rootIt` 🤔

Reply:
> `rootIt` can have attributes instead of being an `emptyIterator()` if there was an insertion before the iteration of `ownAttrsIt` is over. (which is rare)

That's a good point. I guess it's not worth it.